pkg/controller/container-runtime-config/helpers.go

```diff
@@ -511,15 +513,45 @@ func getValidBlockedRegistries(releaseImage string, imgSpec *apicfgv1.ImageSpec)
 	for i, reg := range imgSpec.RegistrySources.BlockedRegistries {
 		// if there is a match, return all the blocked registries except the one that matched and return an error as well
 		if reg == payloadReg {
-			blockedRegs = append(blockedRegs, imgSpec.RegistrySources.BlockedRegistries[i+1:]...)
-			return blockedRegs, fmt.Errorf("error adding %q to blocked registries, cannot block the registry being used by the payload", payloadReg)
+			// If the payload registry doesn't have mirror rules configured for it, then don't add it to the blocked registries list
+			// TODO maybe? Add an opt-in configuration to ICSP and check if that is set as well
+			if !payloadRegHasMirrorConfig(payloadReg, icspRules, imgSpec.RegistrySources.BlockedRegistries) {
+				blockedRegs = append(blockedRegs, imgSpec.RegistrySources.BlockedRegistries[i+1:]...)
+				return blockedRegs, fmt.Errorf("error adding %q to blocked registries, cannot block the registry being used by the payload", payloadReg)
+			}
+			// Log a warning that we are adding the payload registries to the blocked registries list as there are mirror rules for it
+			glog.Warningf("%q is the payload registry, but will add it to the list of blocked registries as there are mirror rules configured for it", reg)
+
 		}
 		// Was not a match to the registry being used by the payload, so add to valid blocked registries
 		blockedRegs = append(blockedRegs, reg)
 	}
 	return blockedRegs, nil
 }
 
+// payloadRegHasMirrorConfig returns true if the payload registry has mirror rules configured for it
+func payloadRegHasMirrorConfig(payloadReg string, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy, blockedRegs []string) bool {
+	if len(icspRules) == 0 || icspRules == nil {
+		return false
+	}
+	var payloadMirrors []string
+	// Go through the mirror rules configured and check if a source matches the payload registry
+	for _, rule := range icspRules {
+		for _, mirror := range rule.Spec.RepositoryDigestMirrors {
+			if runtimeutils.ScopeIsNestedInsideScope(mirror.Source, payloadReg) {
+				payloadMirrors = mirror.Mirrors
+			}
+		}
+	}
+	// Go through the list of mirrors and check that at least one of the payload mirrors is not blocked
+	for _, mirror := range payloadMirrors {
+		if !common.InSlice(mirror, blockedRegs) {
```

Comment:
Shouldn’t this also be a subscope check?

Actually, to be extra correct, it should use `sysregistriesv2.Registry.PullSourcesFromReference` to remap `releaseRepo`, and then check whether that result is blocked — consider a mirror (`quay.io` → `mirror.example.com`), with a `blockedRegs` entry for `mirror.example.com/openshfit-release-dev` (not an exact match for the host name, not an exact match for the repo, a parent repo of the release repo mirror). 

(That would only be possible when integrated in the first loop, where `Source` is available.)

Reply:
So if I understand correctly, calling `sysregistriesv2.Registry.PullSourcesFromReference` on the `releaseRepo` grabs all the possible places that image can be pulled from? so essentially it grabs all the mirrors for that image? 
And you are suggesting we get all the possible pull sources and then check if any of those pull sources are in the list of blocked registries?

Reply:
> So if I understand correctly, calling `sysregistriesv2.Registry.PullSourcesFromReference` on the `releaseRepo` grabs all the possible places that image can be pulled from? so essentially it grabs all the mirrors for that image?

Yes — the complete appropriate repos.

> And you are suggesting we get all the possible pull sources and then check if any of those pull sources are in the list of blocked registries?

I think we might need to, see that example above. A `ScopeIsNestedInsideScope(mirror, blockedReg)` would not find it.

Reply:
Fixed. Using PullSourcesFromReference now, PTAL